c_src/image/cairo_image_ops.c

```diff
@@ -0,0 +1,66 @@
+#include <arpa/inet.h>
+#include <cairo.h>
+#include <stdlib.h>
+
+#include "cairo_ctx.h"
+#include "comms.h"
+#include "image_ops.h"
+
+uint32_t convert_rgba_to_argb(uint32_t pixel)
+{
+  return ((pixel & 0xFFFFFF00) >> 8) | ((pixel & 0xFF) << 24);
+}
+
+int32_t image_ops_create(void* v_ctx,
+                         uint32_t width, uint32_t height,
+                         void* p_pixels)
+{
+  cairo_format_t format = CAIRO_FORMAT_ARGB32;
+  int stride = cairo_format_stride_for_width(format, width);
+  size_t num_pixels = width * height;
+  uint32_t* rgba_pixels = (uint32_t*)p_pixels;
+  uint32_t* argb_pixels = malloc(num_pixels * sizeof(uint32_t));
+
+  for(size_t i = 0; i < num_pixels; ++i) {
+    argb_pixels[i] = convert_rgba_to_argb(htonl(rgba_pixels[i]));
+  }
+
+  image_data_t* image_data = malloc(sizeof(image_data_t));
+
+  image_data->surface
+    = cairo_image_surface_create_for_data((uint8_t*)argb_pixels,
+                                          format,
+                                          width, height, stride);
+
+  image_data->pattern = cairo_pattern_create_for_surface(image_data->surface);
+
+  static cairo_user_data_key_t dummy_key;
+  cairo_surface_set_user_data(image_data->surface, &dummy_key, argb_pixels, free);
+
+  return (int32_t)image_data;
+}
+
+void image_ops_update(void* v_ctx, uint32_t image_id, void* p_pixels)
+{
+  image_data_t* image_data = (image_data_t*)image_id;
+  uint32_t width = cairo_image_surface_get_width(image_data->surface);
+  uint32_t height = cairo_image_surface_get_height(image_data->surface);
+  size_t num_pixels = width * height;
+  uint32_t* rgba_pixels = (uint32_t*)p_pixels;
+  uint32_t* argb_pixels = (uint32_t*)cairo_image_surface_get_data(image_data->surface);
+
+  for(size_t i = 0; i < num_pixels; ++i) {
+    argb_pixels[i] = convert_rgba_to_argb(htonl(rgba_pixels[i]));
+  }
+
+  cairo_pattern_destroy(image_data->pattern);
+  image_data->pattern = cairo_pattern_create_for_surface(image_data->surface);
+}
+
+void image_ops_delete(void* v_ctx, uint32_t image_id)
+{
+  image_data_t* image_data = (image_data_t*)image_id;
```

Comment:
❓ Umm am I missing something, or does this mean this lib isn't supported in aarch64 systems where pointers are uint64_t in size?

Reply:
Yeah, I'm using the image_id to convert to a pointer and assume 32 bit pointer size. Some work would need to be done to support 64 bit

Reply:
The NVG based code is using a image_id that is 32 bits in size and storing this id in the p_image that is stored in tommy_hash. The original code here (https://github.com/gridpoint-com/scenic_driver_local/blob/fec7808e5d73b1fc36345d6590895c1b96edb5b9/c_src/image/image.c#L208-L243) I only changed to remove NVG specific function where the `image_ops_create` and `image_ops_update` functions now exist. I didn't want to mess with the existing surrounding code to avoid breaking existing functionality and only had a 32 bit value to work with.

Reply:
Yea, I didn't mean to imply anything here. I was just making an observation. I wasn't expecting to make any changes around this code as part of this PR.